src/wallet/wallet.cpp

```diff
@@ -4277,11 +4277,11 @@ std::shared_ptr<CWallet> CWallet::CreateWalletFromFile(const std::string& name,
         //We can't rescan beyond non-pruned blocks, stop and throw an error
         //this might happen if a user uses an old wallet within a pruned node
         // or if he ran -disablewallet for a longer time, then decided to re-enable
-        if (fPruneMode)
-        {
+        if (fPruneMode) {
             CBlockIndex *block = chainActive.Tip();
-            while (block && block->pprev && (block->pprev->nStatus & BLOCK_HAVE_DATA) && block->pprev->nTx > 0 && pindexRescan != block)
+            while (block && block->pprev && !IsBlockPruned(block->pprev) && pindexRescan != block) {
```

Comment:
This is slightly different because `!IsBlockPruned` is true when `BLOCK_HAVE_DATA` but from the immediate code alone `nTx > 0` is not guaranteed. Is it the case that `BLOCK_HAVE_DATA` blocks are guaranteed to have `nTx > 0`?

Reply:
Just noting that I see you mention this in your notes: "Note that nTx > 0? is supposed to always be equivalent to (pindex->nStatus & BLOCK_HAVE_DATA)?."

Reply:
It's not identical - via De Morgan's law:
* IsBlockPruned: `!(nStatus & BLOCK_HAVE_DATA) && nTx > 0`
* !IsBlockPruned: `(nStatus & BLOCK_HAVE_DATA) || nTx <= 0`

Reply:
Hm, you're right. I got the `nTx > 0` part backwards... Eesh.

Reply:
Right, so my point is that, if the `false true ?????` case is relevant, then we need to exclude it. Looks like `CChainState::CheckBlockIndex` excludes that possibility, so I'm tending to agree with you: https://github.com/bitcoin/bitcoin/blob/master/src/validation.cpp#L4545

Reply:
I don't think it's ever possible for that case to happen, so I think `!IsBlockPruned()` does what we want here. I.e. it's not a relevant case. Note that this is a check that goes back in the blockchain and only stops if it encounters a missing block. That will only ever be the case if pruning is enabled, and will only ever be caused by pruning. Non-pruning nodes never have holes in the blockchain data (i.e. if they have block at `height=x`, they also have block at `height=x-1` for `x > 0`).